File: ik-otomasyon/src/services/excel.service.js
'use strict';

const XLSX = require('xlsx');

const HEADER_ALIASES = {
  fullName: ['ad soyad', 'adsoyad', 'çalışan adı', 'personel adı', 'isim soyisim', 'isim', 'ad', 'soyad', 'adı'],
  birthDate: ['doğum tarihi', 'doğumtarihi', 'dogum tarihi', 'dogumtarihi', 'doğum', 'dogum', 'birth', 'birthdate', 'doğum günü'],
  hireDate: ['işe giriş tarihi', 'işegiriş tarihi', 'ise giris tarihi', 'işe başlama tarihi', 'işebaşlama tarihi', 'işe giriş', 'başlangıç tarihi', 'başlangıç', 'hire', 'start date'],
};

function normalize(value) {
  return String(value ?? '').trim().toLocaleLowerCase('tr-TR').replace(/\s+/g, ' ');
}

function findColumn(headers, aliases) {
  // Try exact alias match first
  const normalized = headers.map(normalize);
  for (const alias of aliases) {
    const idx = normalized.findIndex((h) => h === normalize(alias));
    if (idx >= 0) return idx;
  }
  // Otherwise try substring/keyword matching
  for (const alias of aliases) {
    const key = normalize(alias);
    const idx = normalized.findIndex((h) => h.includes(key));
    if (idx >= 0) return idx;
  }
  return -1;
}

function findColumnByKeywords(headers, keywords) {
  const normalized = headers.map(normalize);
  for (const key of keywords) {
    const k = normalize(key);
    const idx = normalized.findIndex((h) => h.includes(k));
    if (idx >= 0) return idx;
  }
  return -1;
}

function parseExcelDate(value) {
  if (!value) return null;
  if (value instanceof Date && !Number.isNaN(value.valueOf())) return value;
  if (typeof value === 'number') {
    const parts = XLSX.SSF.parse_date_code(value);
    return parts ? new Date(parts.y, parts.m - 1, parts.d, 12) : null;
  }
  const text = String(value).trim();
  const tr = text.match(/^(\d{1,2})[./-](\d{1,2})[./-](\d{4})$/);
  if (tr) return new Date(Number(tr[3]), Number(tr[2]) - 1, Number(tr[1]), 12);
  const parsed = new Date(text);
  return Number.isNaN(parsed.valueOf()) ? null : parsed;
}

/** Excel satırlarını değiştirmeden tablo için ham veri + otomasyon alanları döndürür. */
function readEmployees(buffer) {
  const workbook = XLSX.read(buffer, { type: 'buffer', cellDates: true, cellNF: true });
  const sheet = workbook.Sheets[workbook.SheetNames[0]];
  if (!sheet) throw new Error('Excel dosyasında okunabilir bir sayfa yok.');

  const matrix = XLSX.utils.sheet_to_json(sheet, { header: 1, defval: '', raw: true });
  const headers = (matrix[0] || []).map((value) => String(value ?? '').trim());
  const fullNameColumn = findColumn(headers, HEADER_ALIASES.fullName);
  const birthDateColumn = findColumn(headers, HEADER_ALIASES.birthDate);
  const hireDateColumn = findColumn(headers, HEADER_ALIASES.hireDate);

  // Try to detect split name columns (e.g., 'Ad' + 'Soyad') if fullName not found
  let adColumn = -1;
  let soyadColumn = -1;
  if (fullNameColumn < 0) {
    adColumn = findColumnByKeywords(headers, ['ad', 'isim', 'adı']);
    soyadColumn = findColumnByKeywords(headers, ['soyad', 'soyadı', 'soy isim']);
  }

  const rows = matrix.slice(1).filter((row) => row.some((value) => String(value ?? '').trim() !== ''));
  const employees = rows.map((row, index) => {
    let fullName = '';
    if (fullNameColumn >= 0) fullName = String(row[fullNameColumn] ?? '').trim();
    else if (adColumn >= 0 || soyadColumn >= 0) {
      const a = String(row[adColumn] ?? '').trim();
      const s = String(row[soyadColumn] ?? '').trim();
      fullName = [a, s].filter(Boolean).join(' ').trim();
    } else {
      // fallback to first non-empty cell in the row
      const first = row.find((c) => String(c ?? '').trim() !== '');
      fullName = String(first ?? '').trim();
    }

    return {
      rowNumber: index + 2,
      values: headers.map((_, column) => row[column] ?? ''),
      fullName,
      birthDate: parseExcelDate(row[birthDateColumn]),
      hireDate: parseExcelDate(row[hireDateColumn]),
    };
  }).filter((employee) => employee.fullName);

  const missing = [];
  if (fullNameColumn < 0 && adColumn < 0 && soyadColumn < 0) missing.push('Ad Soyad');
  if (birthDateColumn < 0) missing.push('Doğum Tarihi');
  if (hireDateColumn < 0) missing.push('İşe Giriş Tarihi');

  return { headers, employees, missing };
}

module.exports = { readEmployees };

File: office-portal/apps/hr/services/excel_service.py
"""
Excel/CSV okuyucu — ik-otomasyon/src/services/excel.service.js'in Python portu.

.xlsx ve .csv tam destekli. Eski .xls (binary) formati bu portta
desteklenmez; kullaniciya acik bir hata mesaji gosterilir (bkz. views.py).
"""
from __future__ import annotations

import csv
import io
import re
from datetime import date, datetime

from openpyxl import load_workbook

HEADER_ALIASES = {
    "full_name": ["ad soyad", "adsoyad", "çalışan adı", "personel adı", "isim soyisim", "isim", "ad", "soyad", "adı"],
    "birth_date": ["doğum tarihi", "doğumtarihi", "dogum tarihi", "dogumtarihi", "doğum", "dogum", "birth", "birthdate", "doğum günü"],
    "hire_date": ["işe giriş tarihi", "işegiriş tarihi", "ise giris tarihi", "işe başlama tarihi", "işebaşlama tarihi", "işe giriş", "başlangıç tarihi", "başlangıç", "hire", "start date"],
}


def _normalize(value) -> str:
    return re.sub(r"\s+", " ", str(value or "").strip().lower())


def _find_column(headers, aliases):
    normalized = [_normalize(h) for h in headers]
    for alias in aliases:
        key = _normalize(alias)
        if key in normalized:
            return normalized.index(key)
    for alias in aliases:
        key = _normalize(alias)
        for i, h in enumerate(normalized):
            if key in h:
                return i
    return -1


def _parse_date(value):
    if value is None or value == "":
        return None
    if isinstance(value, (date, datetime)):
        return date(value.year, value.month, value.day)
    text = str(value).strip()
    match = re.match(r"^(\d{1,2})[./-](\d{1,2})[./-](\d{4})$", text)
    if match:
        day, month, year = int(match.group(1)), int(match.group(2)), int(match.group(3))
        try:
            return date(year, month, day)
        except ValueError:
            return None
    try:
        return datetime.fromisoformat(text).date()
    except ValueError:
        return None


def read_employees(uploaded_file):
    """
    Yuklenen dosyayi okur.

    @returns {"headers": [...], "employees": [...], "missing": [...]}
      employees[i] = {"row_number", "values", "full_name", "birth_date", "hire_date"}
      (birth_date/hire_date "YYYY-MM-DD" string ya da None)
    """
    name = (uploaded_file.name or "").lower()

    if name.endswith(".csv"):
        text = uploaded_file.read().decode("utf-8-sig")
        reader = csv.reader(io.StringIO(text))
        matrix = list(reader)
    elif name.endswith(".xlsx"):
        wb = load_workbook(uploaded_file, data_only=True)
        ws = wb.worksheets[0]
        matrix = [[cell.value for cell in row] for row in ws.iter_rows()]
    else:
        raise ValueError("Bu dosya türü desteklenmiyor. Lütfen .xlsx veya .csv dosyası seçin.")

    if not matrix:
        raise ValueError("Dosya boş görünüyor.")

    headers = [str(h or "").strip() for h in matrix[0]]
    full_name_col = _find_column(headers, HEADER_ALIASES["full_name"])
    birth_col = _find_column(headers, HEADER_ALIASES["birth_date"])
    hire_col = _find_column(headers, HEADER_ALIASES["hire_date"])

    employees = []
    for index, row in enumerate(matrix[1:]):
        if not any(str(v or "").strip() for v in row):
            continue
        full_name = str(row[full_name_col]).strip() if 0 <= full_name_col < len(row) and row[full_name_col] else ""
        if not full_name:
            continue

        birth_date = _parse_date(row[birth_col]) if 0 <= birth_col < len(row) else None
        hire_date = _parse_date(row[hire_col]) if 0 <= hire_col < len(row) else None

        employees.append({
            "row_number": index + 2,
            "values": [row[i] if i < len(row) else "" for i in range(len(headers))],
            "full_name": full_name,
            "birth_date": birth_date.isoformat() if birth_date else None,
            "hire_date": hire_date.isoformat() if hire_date else None,
        })

    missing = []
    if full_name_col < 0:
        missing.append("Ad Soyad")
    if birth_col < 0:
        missing.append("Doğum Tarihi")
    if hire_col < 0:
        missing.append("İşe Giriş Tarihi")

    return {"headers": headers, "employees": employees, "missing": missing}
